src/Lean/Meta/Tactic/AC/Main.lean

```diff
@@ -0,0 +1,337 @@
+    | definitelyNormalized id => mkAppM ``var #[Lean.mkNatLit $ varMap id]
+    | unnormalized _ e => convert varMap e
+    | maybeNormalized _ l r => do mkAppM ``op #[←convert varMap l, ←convert varMap r]
+
+  convertType (op : Lean.Expr) : NormalizedExpr → M Lean.Expr
+    | definitelyNormalized id => do return (←get).exprIdsReverse[id]
+    | unnormalized _ e => convertType op e
+    | maybeNormalized _ l r => do mkAppM' op #[←convertType op l, ←convertType op r]
+
+inductive ProofStrategy
+  | ac_rfl (lhs rhs : NormalizedExpr)
+  | rfl (tgt : Expr)
+  | norm (e : NormalizedExpr)
+
+def pickStrategy (e : Expr) : M ProofStrategy := do
+  match e with
+  | eq l r =>
+    match ←findUnnormalizedOperator l with
+    | lhs@(unnormalized _ _) => return ProofStrategy.norm lhs
+    | lhs@(maybeNormalized _ _ _) =>
+      match ←findUnnormalizedOperator r with
+      | rhs@(unnormalized _ _) => return ProofStrategy.norm rhs
+      | rhs => return ProofStrategy.ac_rfl lhs rhs
+    | lhs@(definitelyNormalized _) =>
+      match ←findUnnormalizedOperator r with
+      | rhs@(unnormalized _ _) => return ProofStrategy.norm rhs
+      | rhs@(maybeNormalized _ _ _) => return ProofStrategy.ac_rfl lhs rhs
+      | rhs@(definitelyNormalized _) => return ProofStrategy.rfl l
+  | e => return ProofStrategy.norm $ ←findUnnormalizedOperator e
+
+def addAcEq (mvarId : MVarId) (e : NormalizedExpr) (target : Expr) : M MVarId := do
+  let (proof, ty) ← buildProof e (←e.norm)
+  let goal ← withLocalDeclD `h_ac ty fun h_ac =>
+    mkForallFVars #[h_ac] target
+  let goal ← mkFreshExprMVar goal
+  assignExprMVar mvarId (mkApp goal proof)
+  return goal.mvarId!
+
+partial def rewriteUnnormalized (mvarId : MVarId) : M Unit :=
+  withMVarContext mvarId do
+  let target ← getMVarType mvarId
+  match ←pickStrategy target with
+  | ProofStrategy.ac_rfl lhs rhs =>
+    trace[Meta.AC] "picking ac_rfl strategy {MessageData.ofGoal mvarId}"
+    try
+      let (proof, ty) ← buildProof lhs rhs
+      if ←isDefEq target ty then
+        assignExprMVar mvarId proof
+      else throwError ""
+    catch _ => throwError "cannot synthesize proof:\n{MessageData.ofGoal mvarId}"
+  | ProofStrategy.rfl tgt =>
+    trace[Meta.AC] "picking rfl strategy {MessageData.ofGoal mvarId}"
+    assignExprMVar mvarId (←mkAppM ``Eq.refl #[tgt])
+  | ProofStrategy.norm (definitelyNormalized _) => throwError "no unnormalized operators found"
+  | ProofStrategy.norm e =>
+    trace[Meta.AC] "picking norm strategy {MessageData.ofGoal mvarId}"
+    let mvarId ← addAcEq mvarId e target
+    let (h_ac, mvarId) ← intro mvarId `h_ac
+    let res ← rewrite mvarId (←getMVarType mvarId) (mkFVar h_ac)
+    let [] := res.mvarIds | throwError "no meta variables expected after rewrite"
+    let mvarId ← replaceTargetEq mvarId res.eNew res.eqProof
+    let mvarId ← clear mvarId h_ac
+    rewriteUnnormalized mvarId
+
+syntax (name := ac_refl) "ac_refl " : tactic
```

Comment:
The syntax declaration should be at `Init/Notation`. Otherwise, users will have to import Lean to use the tactic.

Reply:
done

Reply:
> -- Binders also trigger invalid proofs
> theorem ex2 (n : Nat) : (fun x => n + x) = (fun x => x + n) := by
>   ac_refl
> ```

Most of those were relatively minor. The binder one is harder. I can't just universally quantify over the bound variable, because if it gets applied too much, it could easily cause an oscillation (or at least unnormalize an operator).

The rewrite has to only happen in context. Maybe I can rewrite the bound variable and then wrap it a few times in a `congrArg` or something to only rewrite in context here.

Reply:
The solution is to use `Simp.main`. It will apply all the congruence lemmas and `funext` for us.